### PR TITLE
Colors tests

### PR DESCRIPTION
**Pull request by Kully**
@theengineear 

All tests are passing locally. The plan is to slowly replace the code in `tools` by the methods in `colors`. A renaming of some of the preexisting methods in `colors.py` will also be in order, but only after the code in `tools` is replaced first.

### CODE REVIEW

File: plotly/colors.py [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+    Takes color(s) and a function and applies the function on the color(s)
+
+    In particular, this function identifies whether the given color object
+    is an iterable or not and applies the given color-parsing function to
+    the color or iterable of colors. If given an iterable, it will only be
+    able to work with it if all items in the iterable are of the same type
+    - rgb string, hex string or tuple
+
+    """
+    from numbers import Number
```
Comment (theengineear) [inline plotly/colors.py]: 🐄 imports go at the top of the module.

File: plotly/colors.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+        else:
+            new_color_list = [function(item) for item in colors]
+            return new_color_list
+
+
+def validate_colors(colors):
+    """
+    Validates color(s) and returns an error for invalid colors
+    """
+    from numbers import Number
```
Comment (theengineear) [inline plotly/colors.py]: 🐄 imports a the top of the module

File: plotly/colors.py [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+def convert_colors_to_same_type(colors, colortype='rgb'):
+    """
+    Converts color(s) to the specified color type
+
+    Takes a single color or an iterable of colors and outputs a list of the
+    color(s) converted all to an rgb or tuple color type. If colors is a
+    Plotly Scale name then the cooresponding colorscale will be outputted and
+    colortype will not be applicable
+    """
+    from numbers import Number
```
Comment (theengineear) [inline plotly/colors.py]: 🐄

File: plotly/colors.py [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+            )
+
+    elif isinstance(colors, tuple):
+        if isinstance(colors[0], Number):
+            colors_list = [colors]
+        else:
+            colors_list = list(colors)
+
+    if isinstance(colors, dict):
+        for color in colors.values():
```
Comment (theengineear) [inline plotly/colors.py]: 🐄 `colors_list.extend(colors.values())`?

File: plotly/colors.py [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+        color_tuples.append(new_tuple)
+
+    return color_tuples
+
+
+def label_rgb(colors):
+    """
+    Takes tuple (a, b, c) and returns an rgb color 'rgb(a, b, c)'
+    """
+    return ('rgb(%s, %s, %s)' % (colors[0], colors[1], colors[2]))
```
Comment (theengineear) [inline plotly/colors.py]: Can you just `return u'rgb(%s, %s, %s)' % (colors[0], colors[1], colors[2])` Will that aid your test failure?

File: plotly/colors.py [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+
+def unlabel_rgb(colors):
+    """
+    Takes rgb color(s) 'rgb(a, b, c)' and returns tuple(s) (a, b, c)
+
+    This function takes either an 'rgb(a, b, c)' color or a list of
+    such colors and returns the color tuples in tuple(s) (a, b, c)
+
+    """
+    str_vals = ''
```
Comment (theengineear) [inline plotly/colors.py]: 🐄 could you just use a regex?

http://stackoverflow.com/questions/23205850/regular-expression-help-in-python-looking-for-an-expression-for-rgb-values

File: plotly/colors.py (hunk 5 above)
Comment (Kully) [inline plotly/colors.py]: > Can you just return u'rgb(%s, %s, %s)' % (colors[0], colors[1], colors[2]) Will that aid your test failure?

Does this just return a tuple? I'm not familiar with the notation.

Comment (theengineear) [inline plotly/colors.py]: huh? It's the same thing you had, just with a `u` in front of it for `unicode`, right? I also just removed the unnecessary parentheses around the expression.

File: plotly/colors.py (hunk 6 above)
Comment (Kully) [inline plotly/colors.py]: > 🐄 could you just use a regex?

This is for unlabel right?

File: plotly/tests/test_optional/test_figure_factory.py [hunk 7]
```diff
@@ -842,72 +842,102 @@ def test_trisurf_all_args(self):
...
-            'zerolinecolor': 'rgb(255, 255, 255)'},
-            'zaxis': {'backgroundcolor': 'rgb(230, 230, 230)',
-            'gridcolor': 'rgb(255, 255, 255)',
-            'showbackground': True,
-            'zerolinecolor': 'rgb(255, 255, 255)'}},
-            'title': 'Trisurf Plot',
-            'width': 800
-        }
+            'data': [{'facecolor': np.array(
+                      [b'rgb(145.0, 96.0, 143.0)',
```
Comment (theengineear) [inline plotly/tests/test_optional/test_figure_factory.py]: @Kully remove the `b`'s from here and the `dtype='|S23'`. You're just being overly-strict in this test. Python 2 and 3 treat `str` differently and will create a `byte` or `unicode` string, respectively.

File: plotly/tests/test_core/temp-plot.html [hunk 8]
```diff
@@ -0,0 +1 @@
+<html><head><meta charset="utf-8" /></head><body><div id="854c6dae-0608-4068-8511-3e7e1e6599a6" style="height: 100%; width: 100%;" class="plotly-graph-div"></div><script type="text/javascript">window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL="https://plot.ly";Plotly.newPlot("854c6dae-0608-4068-8511-3e7e1e6599a6", [{"y": [10, 20, 30], "x": [1, 2, 3], "type": "scatter"}], {"title": "offline plot"}, {"linkText": "Export to plot.ly", "showLink": true})</script><script type="text/javascript">window.removeEventListener("resize");window.addEventListener("resize", function(){Plotly.Plots.resize(document.getElementById("854c6dae-0608-4068-8511-3e7e1e6599a6"));});</script></body></html>
```
Comment (theengineear) [inline plotly/tests/test_core/temp-plot.html]: What's going on with this? Did you mean to commit this?

**Comment by theengineear** [comment]
💃 !

**Pull request merged**